pulpcore/cli/common/generic.py

```diff
@@ -195,9 +210,24 @@ def get_command(self, ctx: click.Context, cmd_name: str) -> Optional[click.Comma
         if isinstance(cmd, (PulpCommand, PulpGroup)):
             if cmd.allowed_with_contexts is not None:
                 if not isinstance(ctx.obj, cmd.allowed_with_contexts):
-                    return None
+                    raise IncompatibleContext(
+                        "The command is not available under the current context."
+                    )
         return cmd
 
+    def list_commands(self, ctx: click.Context) -> List[str]:
+        commands_filtered_by_context = []
+        for subcommand in super().list_commands(ctx):
+            try:
+                cmd = self.get_command(ctx, subcommand)
+            except IncompatibleContext:
+                continue
+            else:
+                if cmd is not None:
+                    commands_filtered_by_context.append(subcommand)
+
+        return commands_filtered_by_context
```

Comment:
The original looks like this: https://github.com/pallets/click/blob/cba52fa76135af2edf46c154203b47106f898eb3/src/click/core.py#L1938
What if we say
```python
return sorted([name for name, cmd in self.commands.items() if not hasattr(cmd, "allowed_with_contexts") or ctx.obj in cmd.allowed_with_contexts])
```

I think it is less back and forth between the command object and it's name.

Reply:
In the end, it looks similar to the `get_command` method. Do you find it better?

Reply:
Well they are similar in nature. But the fact that `list_commands` returns the names only makes it a bit harder.